src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/InteropHelpers.cs

```diff
@@ -598,6 +599,33 @@ internal static unsafe void CleanupVariant(IntPtr pDstNativeVariant)
 #endif
         }
 
+        public static unsafe ICustomMarshaler InitializeCustomMarshaller(RuntimeTypeHandle pParameterType, RuntimeTypeHandle pMarshallerType, string cookie, delegate*<string, object> getInstanceMethod)
+        {
+            if (pMarshallerType.IsGenericTypeDefinition())
+            {
+                throw new TypeLoadException();
+            }
+
+            if (pParameterType.ToEETypePtr().IsPointer || pParameterType.IsValueType())
+            {
+                throw new MarshalDirectiveException();
+            }
+
+            if (getInstanceMethod == null)
+            {
+                throw new ApplicationException();
+            }
```

Comment:
We have some error handling in the compiler and some here. I would move all of it to the compiler. It is ok that the errors won't match CoreCLR exactly.